### ISSUE TITLE
Logs not being properly deleted on server

### ISSUE THREAD

**Issue by FinnIckler** [thewca/worldcubeassociation.org#7405]
**Describe the bug**
Ruby will periodically reaper the production logs if they get too big and gzip them. But does not seem to close the files in the system. When running `sudo lsof -nP | grep deleted` you can see that these 10GB files can not be freed by the server as they are still referenced by the ruby processes in some way, even though they are deleted. This will slowly fill up the disk 10GB at a time. This bug must have been here for years, but I was just mitigating it by increasing the disk size. The files can be deleted by running `: > "/proc/$pid/fd/$fd"` or by closing the ruby processes that "use" them.

**To Reproduce**
1. Wait
2. Run `sudo lsof -nP | grep deleted` on Production

**Expected behavior**
When the logs get deleted, they should be deleted completely.

**Screenshot**
![memory leak](https://user-images.githubusercontent.com/10168980/204604777-b4984295-794b-48ba-b1f6-4383f58d5b1b.JPG)

I have not found anything with basic googling about this, but we should fix this asap, as otherwise the website will break after a couple of months of running due to it not being able to create cache fragments. As the screenshot mentions three files reaper.rb, agent_thr(???) and timeout.rb it could be that the process that is suppose to delete the files and close them timeouts, but I have no other leads.

Log Rotation was added in https://github.com/thewca/worldcubeassociation.org/commit/e8eed3acfdf5902565d6ade46c767d9440f9f4db , but I don't think this works correctly as we definitely always have logs over 512MB. Would love some help on this ,maybe even from @jfly or @viroulep, because this must have worked at some point.

**Comment by jfly** [comment]
I would *guess* that logrotation used to work, or maybe we never noticed because we were deployed changes so regularly that would cause our unicorn workers to reopen the log file?

I don't remember where exactly I got that postrotate command, but sending a USR1 *does* seem like the right thing to do. From https://yhbt.net/unicorn/SIGNALS.html:

> USR1 - reopen all logs owned by the master and all workers See Unicorn::Util.reopen_logs for what is considered a log.

That long list of ruby processes definitely looks suspicious. Are you able to create more of them by sending USR1's to the unicorn master process? Does `strace` give you any clue into what they're getting stuck doing?

**Comment by FinnIckler** [comment]
I can not create any more orphaned files by running 
```
[ ! -f /home/cubing/worldcubeassociation.org/WcaOnRails/pids/unicorn.pid ] || kill -USR1 `cat /home/cubing/worldcubeassociation.org/WcaOnRails/pids/unicorn.pid`
```
(at least on staging, but that shouldn't make a difference).

For my understanding: why are we restarting unicorn when we are rotating the rails logs not the unicorn logs?

Log rotation works somewhat, it zips files every 24h, but all the files are 10-12GB instead of the 512MB specified which makes me think that our configuration is at least partially ignored.
I tried `strace`-ing into the processes and found out these are all active, but can't make sense out of the output.

**Comment by jfly** [comment]
> For my understanding: why are we restarting unicorn when we are rotating the rails logs not the unicorn logs?

Ah, good point. Where do we configure rails to write to these `production.log` files? Is that just the default rails behavior? I was assuming that rails was just logging to STDOUT, and [unicorn was redirecting stdout and stderr to a file](https://github.com/thewca/worldcubeassociation.org/blob/ab07bbeb5c6894266c6dc35c648d3eeb48cad0dc/WcaOnRails/config/unicorn.rb#L19-L20), but I don't think that's the case. The [rails 7.0.4 docs say](https://guides.rubyonrails.org/configuring.html#config-logger):

> It defaults to an instance of ActiveSupport::TaggedLogging that wraps an instance of ActiveSupport::Logger which outputs a log to the log/ directory

So here's what I think is happening right now:

1. Things are running happily. There's a `production.log` file with inode 42.
2. Logrotate runs and decides to do a logrotate.
3. Logrotate renames the current `production.log` file to something like `production.log.1`. Note: at this point, there is no longer a `production.log`, and the `production.log.1` file has inode 42, which our various rails processes are still writing to.
4. Logrotate sends a `USR1` to unicorn. Unicorn master and the various unicorn workers all execute this very interesting [`self.reopen_files` method](https://github.com/defunkt/unicorn/blob/d947cb91cf9dbd65d48e497b2605b124571f77ad/lib/unicorn/util.rb#L35) (check it out! it's looking for *all* open files in the ruby process, pretty slick).
   - What I can't figure out is if this `self.reopen_files` method is successfully reopening the log files that were opened by `ActiveSupport::LoggerUnicorn`. It certainly looks like it's trying to. Wild ass guess: this reopen *is* doing something, but it's triggering some weird deadlock/livelock in the worker, and unicorn eventually decides to start a fresh worker? (I have no really good reason to think this other than trying to explain the long list of ruby processes you see).
6. Logrotate exits.

After a whole log of googling, I have yet to find someone explaining clearly how this all should work with Unicorn + Rails. Frustrating results:

- I see a lot of [suggestions to use `copytruncate`](https://stackoverflow.com/a/4883967), which is nice in that it doesn't require applications to close and reopen files. But `copytruncate` *does* mean there's a small window of time where logs can get lost (search for "very small time slice" on https://linux.die.net/man/8/logrotate), so IMO it's not an ideal solution.
- Other folks recommend restarting rails whenever logrotate runs, which feels like an overly heavy hammer, but certainly *would* have the desired effect:
  - ["The reason why is that Rails may still keep pointing to the old log file"](https://stackoverflow.com/questions/4883891/ruby-on-rails-production-log-rotation)
  - ["copytruncate [...] If you don’t use this, you would need to restart your Rails application each time."]
  - (https://gorails.com/guides/rotating-rails-production-logs-with-logrotate)
  - ["copytruncate – this option is required unless you want to restart the rails application server after log rotation"](https://www.botreetechnologies.com/blog/rotate-log-files-in-rails-application-with-logrotate-in-ubuntu/)
- [Here's](https://stackoverflow.com/a/7788080) someone who is either very confused, or Unicorn has changed in the last 10 years since he wrote the answer:
  > Can you use Unicorn? - Unicorn is has built-in support for re-opening all log files in your application via USR1 signal - this allows logrotate to rotate files atomically... - Unicorn keeps track of and restarts it's workers! You can kill the workers after log rotation and Unicorn will restart them, making sure they use the new log file.
  They incorrectly say that sending a USR1 to Unicorn will restart the workers (it doesn't in 2022, at least). They also mention explicitly killing the workers...? I don't know what they're trying to say here.

My best guess is that this really is supposed to "Just Work", and maybe there has been some regression our configuration of Rails doesn't play nicely with Unicorn's `reopen_files` method.

> I can not create any more orphaned files by running [...] (at least on staging, but that shouldn't make a difference).

I wonder if that's because you didn't rename the log file, so we executed [this `next` block in `reopen_logs`](https://github.com/defunkt/unicorn/blob/d947cb91cf9dbd65d48e497b2605b124571f77ad/lib/unicorn/util.rb#L49)? What if you first rename the log file and *then* send USR1 to Unicorn?

**Comment by FinnIckler** [comment]
Running
```
mv production.log production.1.log && [ ! -f /home/cubing/worldcubeassociation.org/WcaOnRails/pids/unicorn.pid ] || kill -USR1 `cat /home/cubing/worldcubeassociation.org/WcaOnRails/pids/unicorn.pid`
```
and then `watch ls -ll` reveals that unicorn is still writing to production.1.log even though it creates a new empty production.log! 

I also haven't found anything where we explicitly configure the name of the log, but settings set [here](https://github.com/thewca/worldcubeassociation.org/blob/595df797a61afe12168f7d56cdb54bb0beb0c415/WcaOnRails/config/environments/production.rb#L123) such as the timestamp and pid are used, so it feels like the rest are defaults.

**Comment by FinnIckler** [comment]
Jeremy and I just looked over this in a call and found out:
- it is not a unicorn process, but a rake job:work process that keeps the files open
- the process writes to the production.log (but it should write to the delayed job Log!)
- this process is not known by supervisor
- restarting supervisor workers does not work because these processes are not associated with super visor (supervisord logs to not mention them)
- looking in the stuck production log the worker seems to be a mailer worker updating the mailer queue

**Comment by jfly** [comment]
> looking in the stuck production log the worker seems to be a mailer worker updating the mailer queue

I don't think we exhaustively checked all the workers, I wouldn't be surprised if both types of workers can get into this weird state.

**Comment by FinnIckler** [comment]
This was semi triaged by reducing the amount of logs by 70%. We need to rethink our logging strategy for our dockerized setup anyway and we are probably just going to send the logs to new relic

**Comment by dunkOnIT** [comment]
@FinnIckler is this still an issue with NewRelic setup?

**Comment by FinnIckler** [comment]
no, because we only log to STDOUT nowadays and that logs to cloudwatch
